src/compiler/checker.ts

```diff
@@ -6386,6 +6386,8 @@ namespace ts {
             // Only narrow when symbol is variable of type any or an object, union, or type parameter type
             if (node && symbol.flags & SymbolFlags.Variable) {
                 if (isTypeAny(type) || type.flags & (TypeFlags.ObjectType | TypeFlags.Union | TypeFlags.TypeParameter)) {
+                    const declaration = getDeclarationOfKind(symbol, SyntaxKind.VariableDeclaration);

```

Comment:
you are walking once to find `top` then another time until you hit `top`. we can put them together, and walk up once, and just stop when you find what would be `top` otherwise.